src/modules/content-s3/content-s3.c

```diff
@@ -27,6 +33,216 @@ struct content_s3 {
+        snprintf (errbuff, eb_size, "failed to parse secret key");
+        goto error;
+    }
+
+    if (!(raw = toml_raw_in (tbl, "access-key-id"))) {
+        saved_errno = errno;
+        free (secret_key);
+        snprintf (errbuff, eb_size, "failed to parse access key");
+        errno = saved_errno;
+        goto error;
+    }
+
+    if (toml_rtos (raw, &access_key)) {
+        saved_errno = errno;
+        free (secret_key);
+        snprintf (errbuff, eb_size, "failed to parse access key");
+        errno = saved_errno;
+        goto error;
+    }
+
+    cfg->secret_key = secret_key;
+    cfg->access_key = access_key;
+
+    return 0;
+
+error:
+    saved_errno = errno;
+    toml_free (tbl);
+    errno = saved_errno;
+    return -1;
+}
+
+static char *hostport (const char *host, const char *port)
+{
+    char *s;
+    if (port == 0) {
+        if (!(s = strdup (host)))
+            return NULL;
+    }
+    else {
+        if (asprintf (&s, "%s:%d", host, port) < 0)
+            return NULL;
+    }
+    return s;
+}
+
+static struct s3_config *parse_config (const flux_conf_t *conf,
+                                       char *errbuff,
+                                       int eb_size)
+{
+    struct s3_config *cfg;
+    flux_conf_error_t error;
+    const char *uri = NULL;
+    const char *bucket = NULL;
+    const char *cred_file = NULL;
+    int is_virtual_host = 0;
+    char tmp[6];
+    struct yuarel yuri;
+    char *cpy = NULL;
+
+    if (!(cfg = calloc (1, sizeof (*cfg))))
+        return NULL;
+
+    cfg->retries = 5;
+    cfg->is_secure = 0;
+
+    if (flux_conf_unpack (conf,
+                          &error,
+                          "{s:{s:s, s:s, s:s, s?:b} !}",
+                          "content-s3",
+                          "credential-file",
+                          &cred_file,
+                          "bucket",
+                          &bucket,
+                          "uri",
+                          &uri,
+                          "virtual-host-style",
+                          &is_virtual_host) < 0) {
+        snprintf(errbuff, eb_size, "%s", error.errbuf);
+        goto error;
+    }
+
+    if (!(cpy = strdup (uri)))
+        goto error;
+
+    if (yuarel_parse (&yuri, cpy) < 0) {
+        free (cpy);
+        snprintf(errbuff, eb_size, "failed to parse uri");
+        errno = EINVAL;
+        goto error;
+    }
+    
+    if (!(cfg->hostname = hostport (yuri.host, yuri.port))) {
+        free (cpy);
+        snprintf(errbuff, eb_size, "failed to form hostname");
+        errno = EINVAL;
+        goto error;
+    }
+
+    if (!(cfg->bucket = strdup (bucket))) {
+        free (cpy);
+        goto error;
+    }
+
+    if (!strncmp (yuri.scheme, "https", 5))
+        cfg->is_secure = 1;
+
+    cfg->is_virtual_host = is_virtual_host;
+
+    if (parse_credentials (cfg, cred_file, errbuff, eb_size))
+        goto error;
+
+    free (cpy);
+    return cfg;
+
+error:
+    s3_config_destroy (cfg);
```

Comment:
you could also just add free(cpy) here (with save/restore of errno), and drop it from the other places you are freeing it.